devicetypes/smartthings/matrix-zdb5100.src/matrix-zdb5100.groovy

```diff
@@ -491,129 +418,120 @@ def off()
-        
-    	sendEvent(name: "color", value: c)
-	} }
-     else{
-    
-    	def hue = value.hue ?: device.currentValue("hue")
-		def saturation = value.saturation ?: device.currentValue("saturation")
-		if(hue == null) hue = 13
-		if(saturation == null) saturation = 13
-		def rgb = huesatToRGB((float)hue, (float)saturation)
-		
+
+    if (value instanceof String) {
+        if ( (value.contains('#') && value.length() == 7) || value.length() == 6) {    //if (value.hex)
+            def c = colorUtil.hexToRgb(value) // value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) }
+            log.debug("Color: R:${c[0]}, G:${c[1]}, B:${c[2]}")
+            result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:c[0], green:c[1], blue:c[2]), 1).format()
+            result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:c[0], green:c[1], blue:c[2]), 2).format()
+            result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:c[0], green:c[1], blue:c[2]), 3).format()
+            result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:c[0], green:c[1], blue:c[2]), 4).format()
+
+            sendEvent(name: "color", value: c)
+        }
+    } else {
+
+        def hue = value.hue ?: device.currentValue("hue")
+        def saturation = value.saturation ?: device.currentValue("saturation")
+        if (hue == null) hue = 13
+        if (saturation == null) saturation = 13
+        def rgb = huesatToRGB((float)hue, (float)saturation)
+
         result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:rgb[0], green:rgb[1], blue:rgb[2]), 1).format()
         result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:rgb[0], green:rgb[1], blue:rgb[2]), 2).format()
         result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:rgb[0], green:rgb[1], blue:rgb[2]), 3).format()
         result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:rgb[0], green:rgb[1], blue:rgb[2]), 4).format()
-	
-	
-	if (value.hue != null) 
-    	sendEvent(name: "hue", value: value.hue)
-        
-
-        
-	if (value.saturation != null) 
-    	sendEvent(name: "saturation", value: value.saturation)
-        }
-	
-	def lastColor = device.latestValue("color")
-    	log.debug("MATRIX: Lastcolor1(${lastColor})")
+
+
+        if (value.hue != null) sendEvent(name: "hue", value: value.hue)
+
+
+
+        if (value.saturation != null) sendEvent(name: "saturation", value: value.saturation)
+    }
+
+    def lastColor = device.latestValue("color")
+    log.debug("MATRIX: Lastcolor1(${lastColor})")
     log.debug("MATRIX: ENUMM   : (${settings.configBtns})")
     log.debug("MATRIX: ENUMM   : (${settings.configBtns.toString()})")
-    if (settings.configBtns=="Button 1" )
-    {
-     result.remove(1)
-     result.remove(1)
-     result.remove(1)
+    if (settings.configBtns == "Button 1") {
```

Comment:
This is a really weird structure. Why do you add _all_ commands and remove all but the one you actually need? Wouldn't it be easier if you just add one command in each _if_ body?

Reply:
I agree, it does look a bit confusing, but the idea is that if settings.configBtns is null, we want to send commands to all 4 endpoints or to dedicated endpoint if Some button is selected

Reply:
Would something like this will be more appealing ?
`

def setColor(value) {
	
    log.debug("MATRIX: setColor(${value})")
    // enchancedLedControl << zwave.configurationV2.configurationGet(parameterNumber: 10)
    log.debug("MATRIX: enchancedLedControl(${state.enchancedLedControl})")
    def result = []
    def WW=0x10
	
    if (state.enchancedLedControl == 1) {
			WW = 0;
            if (settings.configEnchancedLedBlink?.toInteger())        WW |= 1 << 7;
            if (settings.configEnchancedLedBlinkTurnOff?.toInteger()) WW |= 1 << 6;
            if (settings.configEnchancedLedBlinkTurnOn?.toInteger())  WW |= 1 << 5;
            if (settings.configEnchancedLedBlinkDirect?.toInteger())  WW |= 1 << 4;
            WW += settings.configEnchancedLedBlinkTime?.toInteger();
			log.debug("MATRIX: WW=(${WW})")
			log.debug("MATRIX: WW=" + Integer.toBinaryString(WW))
    }
	
    if (value instanceof String) {
        if ( (value.contains('#') && value.length() == 7) || value.length() == 6) {    //if (value.hex)
            def colors = colorUtil.hexToRgb(value) // value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) }
            log.debug("Color: R:${c[0]}, G:${c[1]}, B:${c[2]}")
            sendEvent(name: "color", value: colors)
        }
		
    } else {
        def hue = value.hue ?: device.currentValue("hue")
        def saturation = value.saturation ?: device.currentValue("saturation")
        if (hue == null) hue = 13
        if (saturation == null) saturation = 13
        def colors = huesatToRGB((float)hue, (float)saturation)
        if (value.hue != null) sendEvent(name: "hue", value: value.hue)
        if (value.saturation != null) sendEvent(name: "saturation", value: value.saturation)
    }

    def lastColor = device.latestValue("color")
    log.debug("MATRIX: Lastcolor1(${lastColor})")
    log.debug("MATRIX: ENUMM   : (${settings.configBtns})")
    log.debug("MATRIX: ENUMM   : (${settings.configBtns.toString()})")
    if (settings.configBtns == "Button 1") {
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 1).format()
    }
	
    if (settings.configBtns == "Button 2") {
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 2).format()
    }
	
    if (settings.configBtns == "Button 3") {
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 3).format()
    }
	
    if (settings.configBtns == "Button 4") {
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 4).format()
    }
	
	else
	{	
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 1).format()
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 2).format()
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 3).format()
		result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 4).format()
	}
	
    return response(delayBetween(result, 200))
}
`

Reply:
Yeah, that's definitely better approach, but you have to correct your _if...else_ statements. In your code, all four commands will be added every time `settings.configBtns` is not equal `"Button 4"`
```
 if (settings.configBtns == "Button 1") {
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 1).format()
} else if (settings.configBtns == "Button 2") {
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 2).format()
} else if (settings.configBtns == "Button 3") {
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 3).format()
} else if (settings.configBtns == "Button 4") {
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 4).format()
} else {	
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 1).format()
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 2).format()
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 3).format()
	result << endpointCmd(zwave.switchColorV3.switchColorSet(warmWhite: WW, red:colors[0], green:colors[1], blue:colors[2]), 4).format()
}
```
You want to have something more like this ↑

Reply:
Yes, you are totally righ!